engine/src/main/java/org/terasology/math/ChunkMath.java

```diff
@@ -64,35 +67,188 @@ public static int calcChunkPosZ(int z, int chunkPowerZ) {
+     * @return the relative block in the chunk
+     * @deprecated This is scheduled for removal in an upcoming version
+     *             method will be replaced with JOML implementation {@link #calcChunkPos(Vector3ic, Vector3ic, org.joml.Vector3i)}.
+     */
+    @Deprecated
     public static Vector3i calcChunkPos(Vector3i pos, Vector3i chunkPower) {
         return calcChunkPos(pos.x, pos.y, pos.z, chunkPower);
     }
 
+    /**
+     *
+     * @param pos
+     * @return
+     * @deprecated This is scheduled for removal in an upcoming version
+     *             method will be replaced with JOML implementation {@link #calcChunkPos(Vector3fc, org.joml.Vector3i)}.
+     */
+    @Deprecated
     public static Vector3i calcChunkPos(Vector3f pos) {
         return calcChunkPos(new Vector3i(pos, RoundingMode.HALF_UP));
     }
 
+    /**
+     *
+     * @param pos
+     * @return
+     * @deprecated This is scheduled for removal in an upcoming version
+     *             method will be replaced with JOML implementation {@link #calcChunkPos(Vector3ic, org.joml.Vector3i)}.
+     */
+    @Deprecated
     public static Vector3i calcChunkPos(Vector3i pos) {
         return calcChunkPos(pos.x, pos.y, pos.z);
     }
 
+    /**
+     *
+     * @param x
+     * @param y
+     * @param z
+     * @return
+     * @deprecated This is scheduled for removal in an upcoming version
+     *             method will be replaced with JOML implementation {@link #calcChunkPos(int, int, int, org.joml.Vector3i)}.
+     */
+    @Deprecated
     public static Vector3i calcChunkPos(int x, int y, int z) {
         return calcChunkPos(x, y, z, ChunkConstants.CHUNK_POWER);
     }
 
+    /**
+     *
+     * @param x
+     * @param y
+     * @param z
+     * @param chunkPower
+     * @return
+     * @deprecated This is scheduled for removal in an upcoming version
+     *             method will be replaced with JOML implementation {@link #calcChunkPos(int, int, int, org.joml.Vector3i)}.
+     */
+    @Deprecated
+    public static Vector3i calcChunkPos(int x, int y, int z, Vector3i chunkPower) {
+        return new Vector3i(calcChunkPosX(x, chunkPower.x), calcChunkPosY(y, chunkPower.y), calcChunkPosZ(z, chunkPower.z));
+    }
+
+    //TODO: can be replaced by region3i iterator
     public static Vector3i[] calcChunkPos(Region3i region) {
         return calcChunkPos(region, ChunkConstants.CHUNK_POWER);
     }
 
-    public static Vector3i calcChunkPos(int x, int y, int z, Vector3i chunkPower) {
-        return new Vector3i(calcChunkPosX(x, chunkPower.x), calcChunkPosY(y, chunkPower.y), calcChunkPosZ(z, chunkPower.z));
+    /**
+     * The position relative of the chunk with the given power ({@link ChunkConstants#POWER_X}, {@link ChunkConstants#POWER_Y}, {@link ChunkConstants#POWER_Z})
+     *
+     * <p>Chunk size is in powers of 2 (2, 4, 8, 16, ...)</p>
+     *
+     * @param pos absolute position of the block
+     * @param dest will hold the result
```

Comment:
Clarify? If `dest` is a param, and it gets used in a calculation, the result of which is returned while also matching `dest` ... why was the calculation needed? 🤔 Methinks that probably isn't accurate then?

Reply:
I write it to a destination variable so we can avoid an extra allocation. we use these functions in quite a few loops so reducing the hit to cache seems worth it.

Reply:
So if that happens in a different method, I still think the documentation here is confusing. Because the logic clearly returns the result of a method call to which `dest` is passed - in the end it may well end up the same thing, but you aren't literally returning `dest` you're returning a calculation based on `dest` ?

If the goal is to avoid an extra allocation then you're achieving that in the other method. It doesn't really seem meaningful to tell the user about that here? Shouldn't you doc instead what `dest` _starts_ as, since that's what it is doing as a param?

Reply:
From what I've seen of joml so far, this seems to be the common idiom. The current value of `dest` is irrelevant, it's just a mutable container for stuffing results in to.

I don't love it, because it feels like one of the bad parts of programming in C, but if the joml devs think it helps I find that all-too-believable.

(...if only Java could stack-allocate vectors. then we could just call functions, return results. wouldn't that be nice.)

Reply:
I might be dense, but then I still don't get it. Why have `dest` at all to begin with, if its value is irrelevant? Doesn't the user ultimately end up using the return value, even if that becomes the same value as `dest` in the end? Or is the user meant to subsequently use the variable they passed in as `dest` as it has changed-via-reference?

In either case the docs fail to enlighten this dummy here what's going on :-)

Reply:
a method that takes `dest` _and_ has a return value does have some redundancy in its signature.

returning void would be unambiguous. I expect the return value is provided for convenience so you can write things like `calculateChunkPos(pos, dest).mul(5)` instead of `calculateChunkPos(pos, dest);    dest.mul(5)`

this is a pattern that's going to show up all over the joml-using matrix and vector code, so I think documentation about the pattern would go in the wiki or the package-level docs of terasology.math, probably with references to things like

* https://github.com/JOML-CI/JOML/wiki/Design
* https://github.com/JOML-CI/JOML/wiki/Anti-Patterns-and-Best-Practices#favor-destination-arguments-over-allocating-new-objects

Reply:
So really the only reason to taking `dest` as a param is just to avoid allocating something new?

`* @param dest value ignored, but set to the same as the return value to avoid an allocation`

?

Reply:
hmm, "set to the same as" makes it sound like dest and the return value are separate-but-equal things. But it's the same object.

mmm. I think I need another lunch before I try to think more about technical writing.